server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -876,14 +876,23 @@ public DiscoveryNodes build() {
                 Objects.requireNonNullElse(minNodeVersion, Version.CURRENT.minimumCompatibilityVersion()),
                 Objects.requireNonNullElse(maxDataNodeCompatibleIndexVersion, IndexVersion.current()),
                 Objects.requireNonNullElse(minSupportedIndexVersion, IndexVersion.MINIMUM_COMPATIBLE),
-                dataNodes.values()
-                    .stream()
-                    .flatMap(n -> n.getRoles().stream())
-                    .map(DiscoveryNodeRole::roleName)
-                    .collect(Collectors.toUnmodifiableSet())
+                computeRolesToNodesMap(dataNodes)
             );
         }
 
+        private static Map<String, Set<String>> computeRolesToNodesMap(final Map<String, DiscoveryNode> dataNodes) {
+            Map<String, Set<String>> rolesToNodes = new HashMap<>();
+            for (var node : dataNodes.values()) {
+                for (var role : node.getRoles()) {
+                    rolesToNodes.computeIfAbsent(role.roleName(), (key) -> new HashSet<>()).add(node.getId());
```

Comment:
nit: I think we only care about roles whose `canContainData()` returns `true`. Or really the roles whose names are in `org.elasticsearch.cluster.routing.allocation.DataTier#ALL_DATA_TIERS` plus the legacy `data` role.

Reply:
The map that's used as the argument to this function is already filtered by `conContainData` and I didn't want to duplicate the check.
https://github.com/elastic/elasticsearch/blob/291ecc5e79f2cf79bb9e2ee1abe9fc7e1b70c923/server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java#L865

Reply:
That filters the _nodes_ but we end up with a map keyed by _role_ many of which are not relevant. For instance if you have a node which has roles `data_hot`, `ingest` and `master` then we'll get all three entries in the resulting map, even though we don't care about the `ingest` and `master` roles.

Reply:
Ahh, gotcha, you're right. I'll change this.